File: BE/Seguridad/GrupoPatenteBE.vb
Imports System.Windows.Forms


Public Class GrupoPatenteBE
    Inherits PatenteAbsBE


    Private setPatentes As New HashSet(Of BE.PatenteAbsBE)

    Public Property Patentes() As HashSet(Of BE.PatenteAbsBE)
        Get
            Return setPatentes
        End Get
        Set(ByVal value As HashSet(Of BE.PatenteAbsBE))
            setPatentes = value
        End Set
    End Property


    Public Overrides Function Permite(formulario As String) As Boolean
        For Each patente In setPatentes
            If patente.Permite(formulario) Then
                Return True
            End If
        Next
        Return False
    End Function

    Public Overrides Function ListarPatentesHojasPermitidas() As HashSet(Of PatenteBE)
        Dim ret As New HashSet(Of PatenteBE)
        For Each patente In setPatentes
            ret.UnionWith(patente.ListarPatentesHojasPermitidas())
        Next
        Return ret
    End Function
End Class

File: BE_2/Seguridad/GrupoPatenteBE.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System.Windows.Forms;


public class GrupoPatenteBE : PatenteAbsBE
{
    private HashSet<BE.PatenteAbsBE> setPatentes = new HashSet<BE.PatenteAbsBE>();

    public HashSet<BE.PatenteAbsBE> Patentes
    {
        get
        {
            return setPatentes;
        }
        set
        {
            setPatentes = value;
        }
    }


    public override bool Permite(string formulario)
    {
        foreach (var patente in setPatentes)
        {
            if (patente.Permite(formulario))
                return true;
        }
        return false;
    }

    public override HashSet<PatenteBE> ListarPatentesHojasPermitidas()
    {
        HashSet<PatenteBE> ret = new HashSet<PatenteBE>();
        foreach (var patente in setPatentes)
            ret.UnionWith(patente.ListarPatentesHojasPermitidas());
        return ret;
    }
}
